testsuite/tests/mgc/test_basic.py

```diff
@@ -0,0 +1,105 @@
+from testsuite.openshift.objects.gateway_api import GatewayLight, HTTPRoute
+
+
+# prerequisites:
+# this module assumes the hub cluster is also a spoke cluster so that everything happens on the only cluster
+#
+# multi-cluster-gateways ns is created and set as openshift["project"]
+# managedclustersetbinding is created in openshift["project"]
+# placement named "local-cluster" is created in openshift["project"] and bound to clusterset
+# gateway class "kuadrant-multi-cluster-gateway-instance-per-cluster" is created
+# openshift2["project"] is set
+#
+# dnspolicies are created and bound to gateways automatically by mgc operator
+# dnspolicies leak at this moment
+
+
+@pytest.fixture(scope="module")
+def mgc_apex_domain(openshift):
+    """Returns preconfigured apex domain"""
+    managed_zone = openshift.do_action("get", ["managedzone", "mgc-dev-mz", "-o", "yaml"], parse_output=True)
+    return managed_zone.model["spec"]["domainName"]
+
+
+@pytest.fixture(scope="module")
+def mgc_hostname(blame, mgc_apex_domain):
+    """Returns domain used for testing"""
+    return f"{blame('mgc')}.{mgc_apex_domain}"
+
+
+@pytest.fixture(scope="module")
+def gateway(request, openshift, blame, mgc_hostname):
+    """Creates and returns configured and ready upstream Gateway"""
+    upstream_gateway = GatewayLight.create_instance(
+        openshift=openshift,
+        name=blame("mgc-gateway"),
+        gateway_class_name="kuadrant-multi-cluster-gateway-instance-per-cluster",
+        hostname=mgc_hostname,
+        placement="local-gateway",
+    )
+    upstream_gateway.commit()
+    request.addfinalizer(upstream_gateway.delete)
+
+    return upstream_gateway.wait_for_ready()
+
+
+def _downstream_gateway_of_gateway(gateway, openshift):
+    # in future this function may be a method of GatewayLight if done properly
+    openshift = openshift.change_project(f"kuadrant-{gateway.namespace()}")
+    upstream_gateway = openshift.do_action("get", ["gateway", gateway.name(), "-o", "yaml"], parse_output=False)
+    upstream_gateway = GatewayLight(string_to_model=upstream_gateway.out(), context=openshift.context)
+    # should be deleted automatically by the mgc operator upon mgc gateway deletion
+    return upstream_gateway
+
+
+@pytest.fixture(scope="module")
+def backend(request, openshift2, blame, label):
+    """Deploys Httpbin backend"""
+    httpbin = Httpbin(openshift2, blame("httpbin"), label)
+    request.addfinalizer(httpbin.delete)
+    httpbin.commit()
+    return httpbin
+
+
+@pytest.fixture(scope="module")
+def http_route(gateway, backend, blame, openshift2):
+    """Creates and returns HTTPRoute bound to the backend and the downstream Gateway"""
+    downstream_gateway = _downstream_gateway_of_gateway(gateway, openshift2)
+
+    route = HTTPRoute.create_instance(
+        openshift2,
+        blame("route"),
+        parent=downstream_gateway,
+        hostname=downstream_gateway.model["spec"]["listeners"][0]["hostname"],
```

Comment:
maybe the hostname could be a property of the gateway class?

Reply:
It depends since a gateway can have multiple listeners thus possibly multiple hostnames. It would simplify the implementation but at a cost of generality. I would rather vote for listeners property if desired. Or a full abstraction of listeners according to the gateway api spec -- `Each listener in a Gateway must have a unique combination of Hostname, Port, and Protocol`

Reply:
Right, but currently, you only request the first one. We can add support to multiple ones later